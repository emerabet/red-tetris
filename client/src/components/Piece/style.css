.nextPiece {
	margin-left: 3px;
}

.shapesRow {
	display: flex;
	flex-direction: row;
	justify-content: space-evenly;
	margin: 10px;
}
