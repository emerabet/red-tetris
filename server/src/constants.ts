export enum Direction {
    Down,
    Left,
    Right,
    Top,
}

export enum From {
    Top,
    Bottom,
}

export enum CellState {
    Empty = 0,
    Locked = -1,
}

export const Z:string = 'Z';
export const S:string = 'S';
export const J:string = 'J';
export const L:string = 'L';
export const T:string = 'T';
export const I:string = 'I';
export const O:string = 'O';

export const SHAPES_Z = [
    [
        [1, 1, 0],
        [0, 1, 1],
        [0, 0, 0],
    ],
    [
        [0, 0, 1],
        [0, 1, 1],
        [0, 1, 0],
    ],
    [
        [0, 0, 0],
        [1, 1, 0],
        [0, 1, 1],
    ],
    [
        [0, 1, 0],
        [1, 1, 0],
        [1, 0, 0],
    ],
];

export const SHAPES_S = [
    [
        [0, 2, 2],
        [2, 2, 0],
        [0, 0, 0],
    ],
    [
        [0, 2, 0],
        [0, 2, 2],
        [0, 0, 2],
    ],
    [
        [0, 0, 0],
        [0, 2, 2],
        [2, 2, 0],
    ],
    [
        [2, 0, 0],
        [2, 2, 0],
        [0, 2, 0],
    ],
];

export const SHAPES_J = [
    [
        [3, 0, 0],
        [3, 3, 3],
        [0, 0, 0],
    ],
    [
        [0, 3, 3],
        [0, 3, 0],
        [0, 3, 0],
    ],
    [
        [0, 0, 0],
        [3, 3, 3],
        [0, 0, 3],
    ],
    [
        [0, 3, 0],
        [0, 3, 0],
        [3, 3, 0],
    ],
];

export const SHAPES_L = [
    [
        [0, 0, 4],
        [4, 4, 4],
        [0, 0, 0],
    ],
    [
        [0, 4, 0],
        [0, 4, 0],
        [0, 4, 4],
    ],
    [
        [0, 0, 0],
        [4, 4, 4],
        [4, 0, 0],
    ],
    [
        [4, 4, 0],
        [0, 4, 0],
        [0, 4, 0],
    ],
];

export const SAHPES_T = [
    [
        [0, 5, 0],
        [5, 5, 5],
        [0, 0, 0],
    ],
    [
        [0, 5, 0],
        [0, 5, 5],
        [0, 5, 0],
    ],
    [
        [0, 0, 0],
        [5, 5, 5],
        [0, 5, 0],
    ],
    [
        [0, 5, 0],
        [5, 5, 0],
        [0, 5, 0],
    ],
];

export const SHAPES_I = [
    [
        [0, 0, 6, 0],
        [0, 0, 6, 0],
        [0, 0, 6, 0],
        [0, 0, 6, 0],
    ],
    [
        [0, 0, 0, 0],
        [6, 6, 6, 6],
        [0, 0, 0, 0],
        [0, 0, 0, 0],
    ],
    [
        [0, 6, 0, 0],
        [0, 6, 0, 0],
        [0, 6, 0, 0],
        [0, 6, 0, 0],
    ],
    [
        [0, 0, 0, 0],
        [0, 0, 0, 0],
        [6, 6, 6, 6],
        [0, 0, 0, 0],
    ],
];

export const SHAPES_O = [
    [
        [7, 7],
        [7, 7],
    ],
];
