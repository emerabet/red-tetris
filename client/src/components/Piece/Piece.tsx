import React, { Component, useState, useEffect } from 'react';
import Row from '../Row/Row';
import Cell from '../Cell';

import './style.css';

interface PieceProps {
  piece: number[][];
}

const Piece: React.SFC<PieceProps> = (props) => {
  return (
    <div className="nextPiece">
      {
        props.piece.map((row, index) => {
          return (
            <Row
              key={`row_${index}`}
              row={row}
              index={index}
            />
          );
        })
      }
    </div>
  );
};

export default Piece;
