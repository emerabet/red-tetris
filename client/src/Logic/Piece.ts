import { Direction, CellState } from './constants';

class Piece {
    private readonly shapeName: string;
    private readonly max: number;
    private readonly positions: number[][][];
    private positionRow: number;
    private positionCol: number;
    private index: number;
    private previous: any;

    constructor(shapeName:string, positions: number[][][]) {
        this.positions = positions;
        this.shapeName = shapeName;
        this.index = 0;
        this.max = positions.length;
        this.positionRow = 0;
        this.positionCol = 3;
        this.previous = null;
        this.snapshotPosition();
    }

    private findFirstRow(): number {
        for (let i = 0; i < this.positions[this.index].length; i += 1) {
            for (let j = 0; j < this.positions[this.index][i].length; j += 1) {
                if (this.positions[this.index][i][j] !== CellState.Empty) {
                    return i;
                }
            }
        }
        return 0;
    }

    private findFirstCol(): number {
        for (let col = 0; col < this.positions[this.index][0].length; col += 1) {
            for (let row = 0; row < this.positions[this.index].length; row += 1) {
                if (this.positions[this.index][row][col] !== CellState.Empty) {
                    return col;
                }
            }
        }
        return 0;
    }

    get shape(): number[][] {
        return this.positions[this.index];
    }

    get row(): number {
        return this.positionRow;
    }

    get col(): number {
        return this.positionCol;
    }

    private snapshotPosition() {
        this.previous = {
            index: this.index,
            row: this.positionRow,
            col: this.positionCol,
        };
    }

    rotate() {
        this.snapshotPosition();
        this.index = (this.index + 1) % this.max;
    }

    rollback() {
        if (this.previous) {
            this.index = this.previous.index;
            this.positionRow = this.previous.row;
            this.positionCol = this.previous.col;
        }
    }

    move(direction:Direction) {
        this.snapshotPosition();
        switch (direction) {
        case Direction.Down: this.positionRow += 1; break;
        case Direction.Left: this.positionCol -= 1; break;
        case Direction.Right: this.positionCol += 1; break;
        }
    }
}

export default Piece;
