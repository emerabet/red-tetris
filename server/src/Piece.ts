import { Direction } from './constants';

class Piece {
    private readonly shapeName: string;
    private readonly max: number;
    private readonly positions: number[][][];
    private positionRow: number;
    private positionCol: number;
    private index: number;

    constructor(shapeName:string, positions: number[][][]) {
        this.positions = positions;
        this.shapeName = shapeName;
        this.index = 0;
        this.max = positions.length;
        this.positionRow = 0;
        this.positionCol = 3;
    }

    get shape(): number[][] {
        return this.positions[this.index];
    }

    get row(): number {
        return this.positionRow;
    }

    get col(): number {
        return this.positionCol;
    }

    rotate() {
        this.index = (this.index + 1) % this.max;
    }

    move(direction:Direction) {
        switch (direction) {
        case Direction.Down: this.positionRow += 1;
        case Direction.Left: this.positionCol -= 1;
        case Direction.Right: this.positionCol += 1;
        }
    }
}

export default Piece;
