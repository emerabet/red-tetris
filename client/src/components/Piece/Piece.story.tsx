import React from 'react';
import { storiesOf } from '@storybook/react';
import Piece from './Piece';
import * as utils from '../../Utils/constants';

storiesOf('Piece', module)
    .add('shapes', () => (
        <>
            <div className="shapesRow">
                <Piece
                    piece={utils.SAHPES_T[0]}
                />
                <Piece
                    piece={utils.SAHPES_T[1]}
                />
                <Piece
                    piece={utils.SAHPES_T[2]}
                />
                <Piece
                    piece={utils.SAHPES_T[3]}
                />
            </div>
            <div className="shapesRow">
                <Piece
                    piece={utils.SHAPES_I[0]}
                />
                <Piece
                    piece={utils.SHAPES_I[1]}
                />
                <Piece
                    piece={utils.SHAPES_I[2]}
                />
                <Piece
                    piece={utils.SHAPES_I[3]}
                />
            </div>
            <div className="shapesRow">
                <Piece
                    piece={utils.SHAPES_J[0]}
                />
                <Piece
                    piece={utils.SHAPES_J[1]}
                />
                <Piece
                    piece={utils.SHAPES_J[2]}
                />
                <Piece
                    piece={utils.SHAPES_J[3]}
                />
            </div>
            <div className="shapesRow">
                <Piece
                    piece={utils.SHAPES_L[0]}
                />
                <Piece
                    piece={utils.SHAPES_L[1]}
                />
                <Piece
                    piece={utils.SHAPES_L[2]}
                />
                <Piece
                    piece={utils.SHAPES_L[3]}
                />
            </div>
            <div className="shapesRow">
                <Piece
                    piece={utils.SHAPES_O[0]}
                />
            </div>
            <div className="shapesRow">
                <Piece
                    piece={utils.SHAPES_S[0]}
                />
                <Piece
                    piece={utils.SHAPES_S[1]}
                />
                <Piece
                    piece={utils.SHAPES_S[2]}
                />
                <Piece
                    piece={utils.SHAPES_S[3]}
                />
            </div>
            <div className="shapesRow">
                <Piece
                    piece={utils.SHAPES_Z[0]}
                />
                <Piece
                    piece={utils.SHAPES_Z[1]}
                />
                <Piece
                    piece={utils.SHAPES_Z[2]}
                />
                <Piece
                    piece={utils.SHAPES_Z[3]}
                />
            </div>
        </>
    ));
