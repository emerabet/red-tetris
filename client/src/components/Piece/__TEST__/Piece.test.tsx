import React from 'react';

import renderer from 'react-test-renderer';
import Piece from '../Piece';
import * as utils from '../../../Utils/constants';

it('renders correctly with defaults', () => {
  const piece = renderer.create(<Piece
    piece={utils.SAHPES_T[0]}
     />).toJSON();
  expect(piece).toMatchSnapshot();
});
