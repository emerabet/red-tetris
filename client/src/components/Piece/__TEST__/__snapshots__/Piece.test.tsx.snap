// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with defaults 1`] = `
<div
  className="nextPiece"
>
  <div
    className="boardRow"
  >
    <div
      className="boardCell cell0"
    />
    <div
      className="boardCell cell5"
    />
    <div
      className="boardCell cell0"
    />
  </div>
  <div
    className="boardRow"
  >
    <div
      className="boardCell cell5"
    />
    <div
      className="boardCell cell5"
    />
    <div
      className="boardCell cell5"
    />
  </div>
  <div
    className="boardRow"
  >
    <div
      className="boardCell cell0"
    />
    <div
      className="boardCell cell0"
    />
    <div
      className="boardCell cell0"
    />
  </div>
</div>
`;
